File: solutions/567.permutation-in-string/permutation-in-string.cpp
class Solution {
public:
    bool checkInclusion(string s1, string s2) {
        int len1 = s1.size(), len2 = s2.size();
        if(len1 > len2) return false;
        vector<int> S(26), V(26);
        for(int i = 0; i < len1; i++) {
            S[s1[i] - 97]++;
            V[s2[i] - 97]++;
        }
        if(S == V) return true;
        for(int i = 0; i + len1 < len2; i++) {
            V[s2[i] - 97]--;
            V[s2[len1 + i] - 97]++;
            if(S == V) return true;
        }
        return false;
    }
};

File: solutions/567.permutation-in-string/permutation-in-string.java
class Solution {
    public boolean checkInclusion(String s1, String s2) {
        int[] count = new int[128];
        for(int i = 0; i < s1.length(); i++) count[s1.charAt(i)]--;
        for(int l = 0, r = 0; r < s2.length(); r++) {
            if (++count[s2.charAt(r)] > 0)
                while(--count[s2.charAt(l++)] != 0) { /* do nothing */}
            else if ((r - l + 1) == s1.length()) return true;
        }
        return s1.length() == 0;
    }
}